PowerEditor/src/WinControls/ProjectPanel/Directory.cpp

```diff
@@ -0,0 +1,454 @@
+bool Directory::containsData(const generic_string& path) const
+{
+	if (_filters.empty())
+		return containsData(path, TEXT("*.*"));
+
+	for (size_t i=0; i<_filters.size(); ++i)
+	{
+		if (containsData(path, _filters[i]))
+			return true;
+	}
+	return false;
+}
+
+// containsData() recursive call
+bool Directory::containsData(const generic_string& path, const generic_string& filter) const
+{
+	generic_string searchPath = path + TEXT("\\") + filter;
+	generic_string childPath;
+	generic_string file;
+
+
+	// first step: check only files. If one is found, it contains data.
+	WIN32_FIND_DATA fd;
+	HANDLE hFind = FindFirstFileEx(searchPath.c_str(), FindExInfoBasic, &fd, FindExSearchNameMatch, NULL, 0);
+
+	if (hFind == INVALID_HANDLE_VALUE)
+		return false;
+
+	while (hFind != INVALID_HANDLE_VALUE)
+	{
+
+		// a directory was found... skip it in this pass
+		if ((fd.dwFileAttributes & FILE_ATTRIBUTE_DIRECTORY) != 0)
+		{
+			if (!FindNextFile(hFind, &fd))
+				return false;
+			continue;
+		}
+		else
+		{
+			// data was found - return true
+			FindClose(hFind);
+			return true;
+		}
+
+	}
+
+	if (hFind != INVALID_HANDLE_VALUE)
+		FindClose(hFind);
+
+	// no file was found in this directory. Now check the sub directories.
+	hFind = FindFirstFileEx(searchPath.c_str(), FindExInfoBasic, &fd, FindExSearchLimitToDirectories, NULL, 0);
+
+	if (hFind == INVALID_HANDLE_VALUE)
+		return false;
+
+	while (hFind != INVALID_HANDLE_VALUE)
+	{
+
+		// a directory was found
+		if ((fd.dwFileAttributes & FILE_ATTRIBUTE_DIRECTORY) != 0)
+		{
+			file = fd.cFileName;
+			// skip . and ..
+			if (file == TEXT(".") || file == TEXT(".."))
+			{
+				if (!FindNextFile(hFind, &fd))
+					break;
+				continue;
+			}
+
+			// check if this child directory contains data; if this is the case, return true
+			childPath = path + TEXT("\\") + file;
+			if (containsData(childPath))
+			{
+				FindClose(hFind);
+				return true;
+			}
+		}
+		else
+		{
+			// data was found - return true. This should never happen, because the files were already checked. Just to be safe.
+			FindClose(hFind);
+			return true;
+		}
+
+		if (!FindNextFile(hFind, &fd))
+			break;
+	}
+
+	if (hFind != INVALID_HANDLE_VALUE)
+		FindClose(hFind);
+
+	return false;
+
+}
+
+bool Directory::writeTimeHasChanged() const
+{
+	FILETIME newFiletime;
+
+	// if the new filetime can not be determined, return true for sure. The exact differences will be tested later.
+	if (!readLastWriteTime(newFiletime))
+		return true;
+
+	// we could successfully get the filetime, but before this it was 0 - definitely a change
+	if (!(_lastWriteTime.dwHighDateTime || _lastWriteTime.dwLowDateTime))
+		return true;
+
+	// compare file times
+	return _lastWriteTime.dwLowDateTime != newFiletime.dwLowDateTime
+		|| _lastWriteTime.dwHighDateTime != newFiletime.dwHighDateTime;
+
+}
+
+void Directory::setFilters(const std::vector<generic_string>& filters, bool autoread)
+{
+	if (_filters != filters)
```

Comment:
It has a clear benefit only when `autoread` is true, because there is also an expensive `read()` in this case. Will change it so, that the check is only performed on `autoread`.